mllib/src/main/scala/org/apache/spark/mllib/stat/test/ADTest.scala

```diff
@@ -0,0 +1,264 @@
+   *            exponential, gumbel, logistic, weibull
+   * @param params optional variable-length argument providing parameters for given distribution,
+   *               otherwise they are estimated from sample but in both cases we adjust critical
+   *               values assuming they were estimated from sample. Providing them is simply a
+   *               convenience to avoid recalculation when the values are already available to
+   *               the user
+   * @return Anderson-Darling test result
+   */
+  def testOneSample(data: RDD[Double], distName: String, params: Double*): ADTestResult = {
+    val n = data.count()
+    val makeDist = initDist(distName, data, n, params.toArray)
+    val localData = data.sortBy(x => x).mapPartitions(calcPartAD(_, makeDist, n)).collect()
+    val s = localData.foldLeft((0.0, 0.0)) { case ((prevStat, prevCt), (rawStat, adj, ct)) =>
+      val adjVal = 2 * prevCt * adj
+      val adjustedStat = rawStat + adjVal
+      val cumCt = prevCt + ct
+      (prevStat + adjustedStat, cumCt)
+    }._1
+    val ADStat = -1 * n - s / n
+    val criticalVals = makeDist().getCVs(n)
+    new ADTestResult(ADStat, criticalVals, NullHypothesis.oneSample.toString)
+  }
+
+
+  /**
+   * Calculate a partition's contribution to the Anderson Darling statistic.
+   * In each partition we calculate 2 values, an unadjusted value that is contributed to the AD
+   * statistic directly, a value that must be adjusted by the number of values in the prior
+   * partition, and a count of the elements in that partition
+   * @param part `Iterator[Double]` a partition of the data sample to be analyzed
+   * @param makeDist `() => ADTheoreticalDist` a function to create a class that extends the
+   *                ADTheoreticalDist trait, which requires various methods, used in creating 1
+   *                object per partition
+   * @param n `Double` the total size of the data sample
+   * @return `Iterator[(Double, Double, Double)]` The first element corresponds to the
+   *        position-independent contribution to the AD statistic, the second is the value that must
+   *        be scaled by the number of elements in prior partitions and the third is the number of
+   *        elements in this partition
+   */
+  def calcPartAD(part: Iterator[Double], makeDist: () => ADTheoreticalDist, n: Double)
+    : Iterator[(Double, Double, Double)] = {
+    val dist = makeDist()
+    val initAcc = (0.0, 0.0, 0.0)
+    val pResult = part.zipWithIndex.foldLeft(initAcc) { case ((prevS, prevC, prevCt), (v, i)) =>
+      val y = dist.cdf(v)
+      val a = math.log(y)
+      val b = math.log(1 - y)
+      val unAdjusted = a * (2 * i + 1) + b * (2 * n - 2 * i - 1)
+      val adjConstant = a - b
+      (prevS + unAdjusted, prevC + adjConstant, prevCt + 1)
+    }
+    Array(pResult).iterator
+  }
+
+  /**
+   * Create a function to produce a distribution per partition.
```

Comment:
Thanks for the read! I've bookmarked this page. Initially, under math3 3.1.1 the RealDistribution was not serializable. And since I used it as a member in the various distributions, I thought these had to be instantiated "on the other side". I've changed this in the latest version, since as you point out, I can just have the distribution directly in testOneSample and the closure takes care of it.